src/misc/coloring.c

```diff
@@ -129,6 +133,136 @@ static igraph_error_t igraph_i_vertex_coloring_greedy_cn(const igraph_t *graph,
+    const igraph_vector_int_t *neighbours
+) {
+    igraph_integer_t nbr_cnt = igraph_vector_int_size(neighbours);
+
+    for (igraph_integer_t nbr_indx = 0; nbr_indx < nbr_cnt; nbr_indx++) {
+        igraph_integer_t nbr = VECTOR(*neighbours)[nbr_indx];
+        if (VECTOR(*colors)[nbr] == color) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+static void igraph_i_dsatur_update_heap(
+    const igraph_adjlist_t *adjlist, igraph_gen2wheap_t *node_degrees_heap, 
+    igraph_vector_int_t *neighbours, igraph_vector_int_t *colors, 
+    igraph_integer_t color
+) {
+    igraph_gen2wheap_delete_max(node_degrees_heap);
+    igraph_color_metadata_t color_metadata; 
+    igraph_integer_t nbr_cnt = igraph_vector_int_size(neighbours);
+    for (igraph_integer_t nbr_indx = 0; nbr_indx < nbr_cnt; nbr_indx++){
+        igraph_integer_t nbr = VECTOR(*neighbours)[nbr_indx];
+        if(!igraph_gen2wheap_has_elem(node_degrees_heap, nbr)){
+            continue;
+        }
+        color_metadata = *((igraph_color_metadata_t*)igraph_gen2wheap_get(node_degrees_heap, nbr));
+        if (!is_color_used_by_neighbour(colors, color, igraph_adjlist_get(adjlist, nbr))) {
+            color_metadata.saturation_degree++;
+        }
+        color_metadata.edge_degree--;
+        igraph_gen2wheap_modify(node_degrees_heap, nbr, &color_metadata);
+
+    }
+}
+
+static igraph_integer_t igraph_i_vector_bool_first_true(igraph_vector_bool_t *bool_vector){
+    for(igraph_integer_t index = 0 ; index<igraph_vector_bool_size(bool_vector) ; index++){
+        if(VECTOR(*bool_vector)[index]){
+            return index;
+        }
+    } 
+    return -1;
+}
+static igraph_error_t igraph_i_dsatur_main_loop(
+    const igraph_t *graph, igraph_vector_int_t *colors,
+    igraph_gen2wheap_t *node_degrees_heap, igraph_adjlist_t *adjlist, 
+    igraph_integer_t vc
+) {
+    igraph_integer_t vertices_colored = 0;
+    igraph_vector_bool_t viable_colors;
+    IGRAPH_VECTOR_BOOL_INIT_FINALLY(&viable_colors, vc);
+    while (vertices_colored < vc) {
+        igraph_integer_t node_to_color = igraph_gen2wheap_max_index(node_degrees_heap);
+        igraph_vector_int_t *neighbours = igraph_adjlist_get(adjlist, node_to_color);
+        igraph_integer_t neighbours_count = igraph_vector_int_size(neighbours);
+        igraph_vector_bool_fill(&viable_colors, true);
+        for(igraph_integer_t neighbor_index = 0 ; neighbor_index<neighbours_count ; neighbor_index++){
+            igraph_integer_t neighbour = VECTOR(*neighbours)[neighbor_index]; 
+            if(VECTOR(*colors)[neighbour] != -1){
+                VECTOR(viable_colors)[VECTOR(*colors)[neighbour]] = false;
+            }
+        }
+        igraph_integer_t color = igraph_i_vector_bool_first_true(&viable_colors);
+        igraph_i_dsatur_update_heap(adjlist, node_degrees_heap, neighbours, colors, color);
+        VECTOR(*colors)[node_to_color] = color;
+
+        vertices_colored++;
+    }
```

Comment:
Due to the use of `viable_colors`, we have an $O(|V|^2)$ complexity. Notice that there are several iterations over `viable_colors`, including even the `igraph_vector_bool_fill()` call.

See my earlier comment on how to select new colours efficiently, https://github.com/igraph/igraph/pull/2284#discussion_r1057946570  You can use `igraph_i_vertex_coloring_greedy_cn()` for reference.

Reply:
I have copied the method from igraph_i_vertex_coloring_greedy_cn (in essence, in practice that function created a vector then sorted it, I just created a set instead). So this is done.